exporter/loadbalancingexporter/trace_exporter.go

```diff
@@ -89,79 +153,126 @@ func (e *traceExporterImp) Shutdown(context.Context) error {
-		}
+	te, ok := exp.(exporter.Traces)
+	if !ok {
+		return fmt.Errorf("unable to export traces, unexpected exporter type: expected exporter.Traces but got %T", exp)
+	}
 
-		start := time.Now()
-		err = te.ConsumeTraces(ctx, td)
-		duration := time.Since(start)
+	start := time.Now()
+	err = te.ConsumeTraces(ctx, td)
+	duration := time.Since(start)
 
-		if err == nil {
-			_ = stats.RecordWithTags(
-				ctx,
-				[]tag.Mutator{tag.Upsert(endpointTagKey, endpoint), successTrueMutator},
-				mBackendLatency.M(duration.Milliseconds()))
-		} else {
-			_ = stats.RecordWithTags(
-				ctx,
-				[]tag.Mutator{tag.Upsert(endpointTagKey, endpoint), successFalseMutator},
-				mBackendLatency.M(duration.Milliseconds()))
-		}
+	if err == nil {
+		_ = stats.RecordWithTags(
+			ctx,
+			[]tag.Mutator{tag.Upsert(endpointTagKey, endpoint), successTrueMutator},
+			mBackendLatency.M(duration.Milliseconds()))
+	} else {
+		_ = stats.RecordWithTags(
+			ctx,
+			[]tag.Mutator{tag.Upsert(endpointTagKey, endpoint), successFalseMutator},
+			mBackendLatency.M(duration.Milliseconds()))
 	}
 	return err
 }
 
-func routingIdentifiersFromTraces(td ptrace.Traces, key routingKey) (map[string]bool, error) {
-	ids := make(map[string]bool)
+func validateNotEmpty(td ptrace.Traces) error {
 	rs := td.ResourceSpans()
 	if rs.Len() == 0 {
-		return nil, errors.New("empty resource spans")
+		return errors.New("empty resource spans")
 	}
+	ils := rs.At(0).ScopeSpans()
+	if ils.Len() == 0 {
+		return errors.New("empty scope spans")
+	}
+	spans := ils.At(0).Spans()
+	if spans.Len() == 0 {
+		return errors.New("empty spans")
+	}
+	return nil
+}
 
+func routeByTraceId(td ptrace.Traces) (map[string][]int, error) {
+	ids := make(map[string][]int)
+	rs := td.ResourceSpans()
+	if rs.Len() == 0 {
+		return nil, errors.New("empty resource spans")
+	}
 	ils := rs.At(0).ScopeSpans()
 	if ils.Len() == 0 {
 		return nil, errors.New("empty scope spans")
 	}
-
 	spans := ils.At(0).Spans()
 	if spans.Len() == 0 {
 		return nil, errors.New("empty spans")
 	}
+	tid := spans.At(0).TraceID()
+	ids[string(tid[:])] = []int{}
+	for i := 0; i < rs.Len(); i++ {
+		ids[string(tid[:])] = append(ids[string(tid[:])], i)
+	}
+	return ids, nil
+}
+
+func routeByResourceAttr(td ptrace.Traces, routeKey string) (map[string][]int, error) {
+	ids := make(map[string][]int)
+	err := validateNotEmpty(td)
+	if err != nil {
+		return nil, err
+	}
 
-	if key == svcRouting {
-		for i := 0; i < rs.Len(); i++ {
-			svc, ok := rs.At(i).Resource().Attributes().Get("service.name")
-			if !ok {
-				return nil, errors.New("unable to get service name")
-			}
-			ids[svc.Str()] = true
+	rs := td.ResourceSpans()
+	for i := 0; i < rs.Len(); i++ {
+		attr, ok := rs.At(i).Resource().Attributes().Get(routeKey)
+		if !ok {
+			// If resource attribute is not found, falls back to  Trace ID routing
+			return routeByTraceId(td)
```

Comment:
What if, let's say I have 2 resource spans in TD, and the first one doesn't have `routeKey`. Still, we should route the second resource span (we have batching processor, so there is a good chance this can happen)

Reply:
This is actually a bit confusing. Once a `ptrace.Trace` batch arrives at this function, the expectation is that the resource spans in the `Trace` are already grouped by routing key/trace ID.

For resource attr that happens in this part:

`SplitTracesByResourceAttr`
```
	for routeKey, routeKeyAttrs := range indicesByAttr {
		for _, indices := range routeKeyAttrs {
			tracesForAttr := ptrace.NewTraces()
			for _, i := range indices {
				rs := rss.At(i)
				rs.CopyTo((tracesForAttr.ResourceSpans().AppendEmpty()))
			}
			result[routeKey] = append(result[routeKey], tracesForAttr)
		}
	}
```

for trace ID it happens in this part:

`SplitTracesByResourceAttr`
```
	for j := 0; j < len(fallbackIndices); j++ {
		t := ptrace.NewTraces()
		rs := rss.At(j)
		rs.CopyTo(t.ResourceSpans().AppendEmpty())
		nt := batchpersignal.SplitTraces(t)
		result["traceId"] = append(result["traceId"], nt...)
	}
```

The resulting `map[string][]int)` here is actually mostly unused. We grab the first key and use it to route all spans in the batch:

`consumeTraces`:
```
	routingIds, err := rf(td)
	if err != nil {
		return err
	}
	var rid string
	for key, _ := range routingIds {
		rid = key
		break
	}
	endpoint := e.loadBalancer.Endpoint([]byte(rid))
	exp, err = e.loadBalancer.Exporter(endpoint)
```

This implementation is closer to how the exporter worked before the service name change:
- https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/9c214be44477ef84daa0770a3f0464fc25eae7cf/exporter/loadbalancingexporter/trace_exporter.go

Alternatively, we could just simplify this further and remove this function entirely. We do already access the value for the routing key in `SplitTracesByResourceAttr`. We could just return a `map[string][string][]ptrace.Traces` where the keys are the routing key, and the attribute values, respectively.

Reply:
Thanks for explaining it :-D Yeah, I got it now.